crates/srs/src/voice_stream.rs

```diff
@@ -163,26 +166,52 @@ async fn recv_updates(
     Ok(())
 }
 
-async fn send_updates(
+/**
+  Sends updates about the client to the server. If `game_source` is set,
+  the position and frequency from the latest received `GameMessage` is used.
+  Otherwise, the parameters set in the `client` struct are used.
+*/
+async fn send_updates<G>(
     client: Client,
     mut sink: SplitSink<Framed<TcpStream, MessagesCodec>, Message>,
-    recv_voice: bool,
-) -> Result<(), anyhow::Error> {
+    game_source: Option<G>,
+) -> Result<(), anyhow::Error>
+    where G: Stream<Item=GameMessage> + Unpin
+{
     // send initial SYNC message
     let sync_msg = create_sync_message(&client);
     sink.send(sync_msg).await?;
 
-    loop {
-        delay_for(Duration::from_secs(5)).await;
+    if let Some(mut game_source) = game_source {
+        let mut last_game_msg = None;
+
+        loop {
+            let delay = delay_for(Duration::from_secs(5));
+            // futures::stream::select().await;
+            match future::select(game_source.next(), delay).await {
+                Either::Left((Some(msg), _)) => {
+                    last_game_msg = Some(msg);
+                }
+                Either::Left((None, _)) => {break;}
+                Either::Right((_, _)) => {
+                    info!("Game message timeout")
```

Comment:
If this is mostly for debugging purposes, we could think about demoting it to

```suggestion
                    debug!("Game message timeout")
```